scripts/release

```diff
@@ -45,9 +43,64 @@ if [ -z "$VERSION" ]; then
   exit 1
 fi
 
+# Determine if the version is a release candidate version
+RELEASE_CANDIDATE=false
+if [[ "$VERSION" =~ ^[0-9]+\.[0-9]+\.[0-9]+-rc\.[0-9]+$ ]]; then
+  RELEASE_CANDIDATE=true
+fi
+
+# Ensure that the target commit is an ancestor of master
+git merge-base --is-ancestor $TARGET_COMMIT master
+if [ $? != 0 ]; then
+  if $RELEASE_CANDIDATE; then
+    # Since this is a release candidate, allow it with a warning and confirmation
+    echo -n "You're attempting to publish a release candidate from a non-master branch. Are you sure you want to do this? (y/N) "
+    read ANSWER
+    if [[ ! "$ANSWER" =~ ^[yY]+$ ]]; then
+      echo 'Aborting...'
+      exit 1
+    fi
+  else
+    # In the general case, don't allow it
+    echo "Invalid target: $TARGET"
+    echo 'Please select a target commit which is an ancestor of master.'
+    exit 1
+  fi
+fi
+
+# Update local tags
+echo -n 'Updating local tags via remote fetch... '
+git fetch origin > /dev/null 2>&1
+echo 'done'
+
+# Perform some release candidate assertions, if applicable
+if $RELEASE_CANDIDATE; then
+  SUFFIXLESS_VERSION=${VERSION%-*}
+  # Assert that this release candidate precedes the associated release
+  if [ "$(git tag --list v$SUFFIXLESS_VERSION)" ]; then
+    echo "Cannot create release candidate $VERSION, as a release for $SUFFIXLESS_VERSION already exists."
+    echo "Release candidates must come before proper releases, so please bump the project version in $PROPERTIES_FILE and try again."
+    exit 1
+  fi
+  # Assert that there exists no available lower RC number (e.g. don't allow 1.2.3-rc.2 if 1.2.3-rc.1 is available)
+  RC_NUMBER=1
+  while true; do
+    RC_VERSION="$SUFFIXLESS_VERSION-rc.$RC_NUMBER"
+    # Break once we've reached the current RC number
+    if [ "$VERSION" = "$RC_VERSION" ]; then break; fi
+    # If this rc number is available,
+    if [ -z "$(git tag --list "v$RC_VERSION")" ]; then
+      echo "Cannot create release candidate $VERSION, as $RC_VERSION doesn't exist."
```

Comment:
Just curious, when we release a rc, the prefix version should be in the future or in the past?
I think the common should be:
say we are now at v0.1 
we are going to v0.2 but have some rc,
so we have
v0.2-rc1
v0.2-rc2
then when everything works,
v0.2 is the released.

We shall have more rc\d  and probably the prefix version string should never happened?

I verified other rc conventions by looking at linux
https://github.com/torvalds/linux/releases

Reply:
I agree with your assessment, that's how it should be. The release script is designed with this pattern in mind, and it fails in the following cases:
- Trying to release `0.1-rc.1` after `0.1` has already been released (line `81`)
- Trying to release `0.1-rc.2` before `0.1-rc.1` has been released (applies for any combination of numbers; line `93`)

The logic you've highlighted is the latter.